Use repository factories

dantleech opened the pull request:
This PR makes the bundle more extensible by introducing the concept of repository factories.

Instead of the service instantaition being hard-coded in the DI extension (and
limited to the 3 drivers that we provide) service instantiation is taken care
of by a dedicated Factory class for each driver.

These factories are tagged using the `cmf_resource.repository_factory` tag.

Note that this PR is backwards compatible.

dantleech commented on Factory/FilesystemFactory.php:
```diff
@@ -0,0 +1,42 @@
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Cmf\Component\Resource\Repository\PhpcrOdmRepository;
+
+class FilesystemFactory extends RepositoryFactoryInterface
```
This is an interface (as-yet unpublished) in the `resource` component.

dantleech commented on Factory/FilesystemFactory.php:
```diff
@@ -0,0 +1,42 @@
+        return [
+            'base_dir' => null,
+            'symlink' => null,
+        ];
```
This method is part of the interface and provides a good comprimise for providing default values / validating keys and performance / decoupling (i.e. using the OptionsResolver).

dantleech commented:
@WouterJ wdyt? I think this is similar to the Symfony SecuityProvider pattern? (or?)

dantleech commented on Registry/RepositoryRegistry.php:
```diff
@@ -0,0 +1,151 @@
+/**
+ * @author Daniel Leech <daniel@dantleech.com>
+ */
+class RepositoryRegistry implements RepositoryRegistryInterface
```
The Registry implementation now accepts Factories.

dantleech commented on Factory/FilesystemFactory.php:
```diff
@@ -0,0 +1,40 @@
+
+class FilesystemFactory implements RepositoryFactoryInterface
+{
+    public function create(array $options)
```
inherirdoc

dantleech commented on Factory/PhpcrFactory.php:
```diff
@@ -0,0 +1,41 @@
+        $this->container = $container;
+    }
+
+    public function create(array $options)
```
inheritdoc

dantleech commented on Factory/PhpcrOdmFactory.php:
```diff
@@ -0,0 +1,41 @@
+        $this->container = $container;
+    }
+
+    public function create(array $options)
```
ihdoc

dantleech commented on Registry/RepositoryRegistry.php:
```diff
@@ -0,0 +1,151 @@
+
+    private $defaultInstanceName;
+
+    public function __construct(
```
dc

dantleech commented on Resources/config/factories.xml:
```diff
@@ -0,0 +1,28 @@
+
+        <service id="cmf_resource.factory.doctrine_phpcr_odm" class="Symfony\Cmf\Bundle\ResourceBundle\Factory\PhpcrOdmFactory">
+            <argument type="service" id="service_container" />
+            <tag name="cmf_resource.repository_factory" alias="doctrine/phpcr-odm" />
```
Previously the values were `doctrine_phpcr_odm` etc. This IS a BC break, revert.

dantleech commented on Resources/config/resource.xml:
```diff
@@ -4,18 +4,16 @@
-            <argument type="collection" />
+        <service id="cmf_resource.registry" class="Symfony\Cmf\Bundle\ResourceBundle\Registry\RepositoryRegistry">
+            <argument>%cmf_resource.repository_configuration%</argument>
+            <argument>default</argument>
```
The default repository should be configurable --

dantleech commented:
@WouterJ @dbu @lsmith77 as described, this PR uses Factories to instantiate the Repository instances (e.g. PhpcrOdmRepository, FilesystemRepository). It retains BC, if no objects I will merge in at least 2 days.

dantleech commented on Tests/Unit/DependencyInjection/ConfigurationTest.php:
```diff
@@ -14,62 +14,79 @@
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Loader\YamlFileLoader;
 
-class ConfigurationTest extends AbstractExtensionConfigurationTestCase
```
I removed the DI testing library -- in general I find it to abstract what is actually going on, which, infact, isn't that much.

dantleech commented on Tests/Unit/DependencyInjection/ConfigurationTest.php:
```diff
@@ -14,62 +14,79 @@
 {
-    protected function getContainerExtension()
+    private $configuration;
+    private $processor;
```
dd

WouterJ commented:
> I think this is similar to the Symfony SecuityProvider pattern? (or?)

As far as I can see now (had a quick look), it's actually quite different. I'll test it out later tonight/tomorrow morning and will give some more detailed feedback (or maybe just an immediate +1 ;))

ElectricMaxxx commented on Resources/config/factories.xml:
```diff
@@ -0,0 +1,28 @@
+            <tag name="cmf_resource.repository_factory" alias="filesystem" />
+        </service>
+
+        <service id="cmf_resource.factory.doctrine_phpcr" class="Symfony\Cmf\Bundle\ResourceBundle\Factory\PhpcrFactory">
```
We do not have to devide services definitions by its persistence layer in this bundle yet, right?

dantleech replied:
This PR means that we don't have to. The reason we separate them elsewhere is so that the service definitions which reference "suggested" dependencies are only loaded when explicitly enabled.

On 13 May 2016 05:32:00 BST, Maximilian Berghoff <notifications@github.com> wrote:
>> @@ -0,0 +1,28 @@
>> +<?xml version="1.0" encoding="UTF-8"?>
>> +<container xmlns="http://symfony.com/schema/dic/services"
>> +    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
>> +    xsi:schemaLocation="http://symfony.com/schema/dic/services
>http://symfony.com/schema/dic/services/services-1.0.xsd">
>> +
>> +    <services>
>> +
>> +        <service id="cmf_resource.factory.filesystem"
>class="Symfony\Cmf\Bundle\ResourceBundle\Factory\FilesystemFactory">
>> +            <tag name="cmf_resource.repository_factory"
>alias="filesystem" />
>> +        </service>
>> +
>> +        <service id="cmf_resource.factory.doctrine_phpcr"
>class="Symfony\Cmf\Bundle\ResourceBundle\Factory\PhpcrFactory">
>
>We do not have to devide services definitions by its persistence layer
>in this bundle yet, right?
>
>---
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/symfony-cmf/resource-bundle/pull/18/files/48e552f9c7e6f6d3844e50994774b400b0a69960#r63132386

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

WouterJ replied:
it'll simplify usage, +1 for including it

dantleech replied:
Including what?

ElectricMaxxx commented on Tests/Unit/DependencyInjection/ConfigurationTest.php:
```diff
@@ -13,63 +13,76 @@
-                            array(
-                                'repository' => 'articles',
-                                'mountpoint' => '/articles',
+                    'stuff' => array(
```
shouldn't we use short array notation here too?

WouterJ commented on Registry/RepositoryRegistry.php:
```diff
@@ -0,0 +1,147 @@
+        return $this->typeMap[$repositoryClass];
+    }
+
+    private function createRepositoryInstance($instanceName)
```
why is this not included in the DI extension and then registered as a method call to `addFactory`? This allows us to fail fast when there is a misconfiguration and it allows the repository factories to use the TreeBuilder of the config component to explicitely define the options that can be configured.

dantleech replied:
You mean create the instances in the DI extension and register them here? I suppose that could be a better solution. Will have a look.

WouterJ replied:
Yes. That's how Security does it as well

WouterJ commented on Registry/RepositoryRegistry.php:
```diff
@@ -0,0 +1,147 @@
+    private $typeMap = [];
+
+    public function __construct(
+        array $configurations
```
no need for multi-line argument listings here

WouterJ commented on Tests/Unit/DependencyInjection/CmfResourceExtensionTest.php:
```diff
@@ -12,75 +12,32 @@
+                    ],
+                ],
+            ],
+        ], $container);
```
I think we should keep using the `array()` syntax until we dropped CMF 1 support (and then, just update StyleCI config to change everything in all maintained branches at once, so we won't have merge conflicts).

dantleech replied:
Yeah, its just my habit now. Style CI will fix it.

dantleech replied:
Although Style CI actually doesn't seem to care. Will change manually.

dantleech replied:
Since this is a `0.x.x` bundle, and this PR updates it to have a minimum PHP version of 5.5 (as with Symfony 3 and the `master` branch of RoutingAuto), what harm could come from using the new syntax? Saying that it seems that Symfony 3 will not adopt the new syntax whilst changes have to be merged from 2.8, so fine with using the old syntax.

WouterJ replied:
Hmm, indeed. We should then use the short syntax everywhere in this bundle. Forgot that this bundle is not yet stable

dantleech commented:
Updated: The registry now actively instantiates all of the configured repositories and is immutable.

Also updated Style CI to force the short array syntax.

dantleech commented:
> The registry now actively instantiates all of the configured repositories and is immutable.

Actually that doesn't work very well because of circular references, will need to try something else.

dantleech commented on Registry/RepositoryRegistry.php:
```diff
@@ -0,0 +1,141 @@
+        $config = array_merge($defaultConfig, $config);
+
+        if ($factory instanceof CompositeFactory) {
+            $factory->setRepositoryRegistry($this);
```
The `CompositeFactory` requires an instance of the `RepositoryRegistry`, which requires an instance of the `CompositeFactory` ... so constructor injection of the registry to the factory will not work.

There are then 2 solutions that I saw: 1. Passing the registry as a second argument to `create` and 2. Doing this.

Rather than create a `RepositoryRegistryAwareInterface` I just type check on the `CompositeFactory` - it is a special case at the moment, if needs require we can add an `aware` interface later.

dantleech replied:
This class will be completely refactored.

dantleech commented:
Updated (again) avoiding the circular dependency.

dantleech commented on composer.json:
```diff
@@ -13,12 +13,10 @@
         "php": "^5.3.9|^7.0",
         "symfony/framework-bundle": "^2.3|3.*",
-        "symfony-cmf/resource": "^1.0"
+        "symfony-cmf/resource": "dev-repository_factory_int"
```
revert this before merging.

WouterJ commented:
What about refactoring this to: (this is what I meant originally)

```php
class FilesystemFactory implements RepositoryFactoryInterface
{
    /**
     * {@inheritdoc}
     */
    public function create(ContainerBuilder $container, array $options)
    {
        if (null === $options['base_dir']) {
            throw new \InvalidArgumentException(
                'The filesystem repository type requires the `base_dir` option to be set.'
            );
        }

        $container->setDefinition('cmf_resource.repository.filesystem', new Definition(
            FilesystemRepository::class,
            [$options['base_dir'], $options['symlink']]
        ));
    }

    /**
     * {@inheritdoc}
     */
    public function addConfiguration(NodeDefinition $node)
    {
        $node->children()
                ->scalarNode('base_dir')->defaultNull()->end()
                ->booleanNode('symlink')->defaultTrue()->end()
            ->end();
    }
}
```

dantleech commented:
@WouterJ this is what Sylius does too (in a slightly different context).

So - we would parse the repository configuration build up the required DI definitions using configurator classes (as in your example) then pass the container and service IDs to the Registry?

The only issue I see here is that we cannot register these "DI" configurators with tags, they would need to be added manually and the source of knowledge about them would reside in this bundle, or?

WouterJ commented:
> So - we would parse the repository configuration build up the required DI definitions using configurator classes (as in your example) then pass the container and service IDs to the Registry?

Yeah, or immediately pass `Reference` instances to the Registry (the drupal guys discovered that this was faster than getting each service ID when needed in the EventDispatcher).

 > The only issue I see here is that we cannot register these "DI" configurators with tags, they would need to be added manually and the source of knowledge about them would reside in this bundle, or?

The SecurityBundle resolves it this way: https://github.com/symfony/security-bundle/blob/master/SecurityBundle.php#L42-L57

dantleech commented:
> The SecurityBundle solves this problem like this

But that couples the implementations to this bundle. My aim with this PR was to allow 3rd-party repositories to be used seamlessly - so that there is no distrinction between the repositories we provide and those which a developer might choose to write/use.

WouterJ commented:
Unless I'm wrong, any bundle can do `->get('security')->addRepositoryFactory(...)` in the `Bundle#build()` method.

dantleech commented:
ah I see.

```php
$extension = $container->getExtension('security');
$extension->addSecurityListenerFactory(new FormLoginFactory());
$extension->addSecurityListenerFactory(new HttpBasicFactory());
```

Looks good then. Will try it.

dantleech commented:
One issue is that the configuration:

We do this:

```yaml
cmf_resource:
    repositories:
        test_repository:
            type: doctrine_phpcr_odm
            basepath: /test
```

We don't do this:

```yaml
cmf_resource:
    repositories:
        doctrine_phpcr_odm:
            name: test_repository
            basepath: /test
```

So we are going to have to parse the configuration before building the configuration (if thats possible) and dynamically build the configuration depending on the configuration ... which sounds implausable.

Otherwise we can switch the configuration format (to something less user friendly imo) or we defer option validation to the `create` method of the factory.

WouterJ commented:
Hmm, indeed. Let's not change the configuration and put option validation in create() instead (maybe use OptionsResolver component?).

dantleech commented:
@WouterJ have made an initial migration and fully implemented the CompsiteFactory. Will continue later this afternoon.

dantleech commented on DependencyInjection/Repository/Factory/CompositeFactory.php:
```diff
@@ -0,0 +1,99 @@
+        $options->setDefault('mounts', []);
+        $options->setAllowedTypes('mounts', [ 'array' ]);
+
+        $options->setNormalizer('mounts', function (Options $options, $value) {
```
Not sure if there is a better place to do this validation.

dantleech commented on DependencyInjection/Repository/Factory/CompositeFactory.php:
```diff
@@ -0,0 +1,99 @@
+     */
+    public function configure(OptionsResolver $options)
+    {
+        // TODO: How to validate? in Normalizer?
```
remove comment.

dantleech commented on DependencyInjection/CmfResourceExtension.php:
```diff
@@ -14,143 +14,92 @@
+                $config = $optionsResolver->resolve($config['options']);
+            } catch (\Exception $e) {
+                throw new InvalidArgumentException(sprintf(
+                    'Invalid configuration for repository "%s"',
```
Maybe slightly controversial - I am wrapping any exceptions thrown by the OptionsResolver in order to indicate on which repository the error occurred.

dantleech commented on DependencyInjection/CmfResourceExtension.php:
```diff
@@ -14,143 +14,92 @@
-                $name,
-                implode('", "', array_keys($options)),
-                implode(', ', $knownOptions)
+        if (isset($this->repositoryFactories[$name])) {
```
Add test.

dantleech replied:
Already done.

dantleech commented:
Only thing to do now is refactor the Registry class - @WouterJ what did you mean by passing the `Reference` instances - can you provide an example? 

Otherwise I am pretty happy with this new approach.

dantleech commented on DependencyInjection/Repository/Factory/DoctrinePhpcrFactory.php:
```diff
@@ -0,0 +1,61 @@
+     * @param mixed            $userProvider
+     * @param mixed            $defaultEntryPoint
+     */
+    public function create(array $options)
```
inheritfdoc

dantleech commented on DependencyInjection/Repository/Factory/DoctrinePhpcrFactory.php:
```diff
@@ -0,0 +1,61 @@
+     *
+     * @return string
+     */
+    public function getName()
```
inheritdoc

dantleech commented on DependencyInjection/Repository/Factory/DoctrinePhpcrFactory.php:
```diff
@@ -0,0 +1,61 @@
+     *
+     * @param OptionsResolver
+     */
+    public function configure(OptionsResolver $options)
```
indc

dantleech commented on DependencyInjection/Repository/Factory/DoctrinePhpcrOdmFactory.php:
```diff
@@ -0,0 +1,61 @@
+     * @param mixed            $userProvider
+     * @param mixed            $defaultEntryPoint
+     */
+    public function create(array $options)
```
all inheritdocs

dantleech commented on DependencyInjection/Repository/Factory/FilesystemFactory.php:
```diff
@@ -0,0 +1,52 @@
+     *
+     * @param OptionsResolver
+     */
+    public function configure(OptionsResolver $options)
```
id

dantleech commented on DependencyInjection/Repository/Factory/RepositoryFactoryInterface.php:
```diff
@@ -0,0 +1,43 @@
+     * @param mixed            $id
+     * @param mixed            $config
+     * @param mixed            $userProvider
+     * @param mixed            $defaultEntryPoint
```
remove params

dantleech commented on Registry/RepositoryRegistry.php:
```diff
@@ -0,0 +1,93 @@
+
+    /**
+     * @param array $serviceMap
+     * @param array $configurations
```
missing param

dantleech commented on Registry/RepositoryRegistry.php:
```diff
@@ -0,0 +1,93 @@
+        ));
+    }
+
+    public function getRepositoryType(ResourceRepository $repository)
```
inheritdoc

dantleech commented on Resources/config/description.xml:
```diff
@@ -0,0 +1,23 @@
+
+        <service id="cmf_resource.registry.container" class="%cmf_resource.registry.container.class%">
+            <argument type="service" id="service_container" />
+            <argument type="collection" />
```
rm this file.

dantleech commented:
Have updated the Registry and fixed everything. If the tests are green this is in a final state. /cc @WouterJ

dantleech commented:
Updated to use symfony 2.6 minimum.

WouterJ commented:
> Updated to use symfony 2.6 minimum.

For CMF 2.0, we decided to go for 2.8+ (or maybe 2.7+ if that's possible without introducing any BC layers). 2.6 reached end of maintaince a long time ago, it's not worth supporting it.

WouterJ commented on DependencyInjection/CmfResourceExtension.php:
```diff
@@ -30,125 +44,70 @@ public function load(array $configs, ContainerBuilder $container)
-        }
-
-        unset($options['mounts']);
+            $serviceId = self::getRepositoryServiceId($repositoryName);
```
this is highly based on conventions. Why not make this part of the Factory (i.e. passing the container builder or by introducing a `getRepositoryServiceId` method?).

WouterJ commented on DependencyInjection/Repository/Factory/CompositeFactory.php:
```diff
@@ -0,0 +1,91 @@
+    public function configure(OptionsResolver $options)
+    {
+        $options->setDefault('mounts', []);
+        $options->setAllowedTypes('mounts', ['array']);
```
This can be just `setAllowedTypes('mounts', 'array')`, but it's not that important.

WouterJ commented on DependencyInjection/Repository/Factory/DoctrinePhpcrFactory.php:
```diff
@@ -0,0 +1,50 @@
+            $options['basedir'],
+        ]);
+
+        return $definition;
```
can be simplified as:

```php
return new Definition(PhpcrRepository::class, [
    new Reference('doctrine_phpcr.session').
    $options['basedir'],
]);
```

WouterJ commented on DependencyInjection/Repository/Factory/FilesystemFactory.php:
```diff
@@ -0,0 +1,50 @@
+     */
+    public function configure(OptionsResolver $options)
+    {
+        $options->setRequired(['base_dir']);
```
`setRequired('base_dir')`

WouterJ commented on DependencyInjection/Repository/Factory/RepositoryFactoryInterface.php:
```diff
@@ -0,0 +1,55 @@
+     * Configure a repository service using the given options.
+     *
+     * @param array $options
+     */
```
missing `@return`

WouterJ commented on composer.json:
```diff
@@ -12,13 +12,11 @@
     "minimum-stability": "dev",
     "require": {
         "php": "^5.3.9|^7.0",
```
should be updated to `^5.5.9|^7.0`

dantleech commented:
Updated. Also updated to Symfony 2.7 as the minumum .. ?

dantleech commented:
@WouterJ feel free to merge, otherwise I will do so later this afternoon.
